src/commonMain/kotlin/io/github/detekt/sarif4k/SarifSerializer.kt

```diff
@@ -15,4 +20,57 @@ object SarifSerializer {
     fun toJson(sarif: SarifSchema210): String = json.encodeToString(sarif)
 
     fun fromJson(json: String): SarifSchema210 = Json.decodeFromString(json)
-}
\ No newline at end of file
+}
+
+// Extracted from https://github.com/Kotlin/kotlinx.serialization/issues/296#issuecomment-1859572541
+private fun Any?.toJsonElement(): JsonElement = when (this) {
+    null -> JsonNull
+    is JsonElement -> this
+    is Map<*, *> -> toJsonElement()
+    is Collection<*> -> toJsonElement()
+    is ByteArray -> this.toList().toJsonElement()
+    is CharArray -> this.toList().toJsonElement()
+    is ShortArray -> this.toList().toJsonElement()
+    is IntArray -> this.toList().toJsonElement()
+    is LongArray -> this.toList().toJsonElement()
+    is FloatArray -> this.toList().toJsonElement()
+    is DoubleArray -> this.toList().toJsonElement()
+    is BooleanArray -> this.toList().toJsonElement()
+    is Array<*> -> toJsonElement()
+    is Boolean -> JsonPrimitive(this)
+    is Number -> JsonPrimitive(this)
+    is String -> JsonPrimitive(this)
+    is Enum<*> -> JsonPrimitive(this.toString())
+    else -> error("Can't serialize unknown type: $this")
+}
+
+internal fun Map<*, *>.toJsonElement(): JsonObject {
+    return JsonObject(this.map { (key, value) -> key as String to value.toJsonElement() }.toMap())
+}
+
+private fun Collection<*>.toJsonElement(): JsonArray {
+    return JsonArray(this.map { it.toJsonElement() })
+}
+
+private fun Array<*>.toJsonElement(): JsonArray {
+    return JsonArray(this.map { it.toJsonElement() })
+}
+
+private fun JsonElement.toMap(): Any? {
+    return when (this) {
+        JsonNull -> null
+        is JsonArray -> this.map { it.toMap() }
+        is JsonObject -> this.toMap()
+        is JsonPrimitive -> if (isString) content else {
+            content
+                .toBooleanStrictOrNull()
+                ?: content.toDoubleOrNull()
+                ?: content.toLongOrNull()
```

Comment:
I think we want to swap these:

```suggestion
                ?: content.toLongOrNull()
                ?: content.toDoubleOrNull()
```

because `123` is a valid floating point:
![image](https://github.com/detekt/sarif4k/assets/2906988/197f94b4-598e-461a-a36f-da5a082db7ce)

which means it'll never be parsed as an integer type.